File: shared/tscanner-common/src/types/params.ts
import z from 'zod';
import { AiExecutionMode } from './enums';

const previousAiIssueSchema = z.object({
  rule: z.string(),
  file: z.string(),
  line: z.number(),
  column: z.number(),
  message: z.string(),
  line_text: z.string().optional(),
});

const scanParamsSchema = z.object({
  root: z.string(),
  config: z.any().optional(),
  config_dir: z.string().optional(),
  branch: z.string().optional(),
  staged: z.boolean().optional(),
  ai_mode: z.nativeEnum(AiExecutionMode).optional(),
  no_cache: z.boolean().optional(),
  previous_ai_issues: z.array(previousAiIssueSchema).optional(),
});

const scanFileParamsSchema = z.object({
  root: z.string(),
  file: z.string(),
});

const scanContentParamsSchema = z.object({
  root: z.string(),
  file: z.string(),
  content: z.string(),
  config: z.any().optional(),
  config_dir: z.string().optional(),
  branch: z.string().optional(),
  uncommitted: z.boolean().optional(),
});

const modifiedLineRangeSchema = z.object({
  startLine: z.number(),
  lineCount: z.number(),
});

export type ScanParams = z.infer<typeof scanParamsSchema>;
export type PreviousAiIssue = z.infer<typeof previousAiIssueSchema>;
export type ScanFileParams = z.infer<typeof scanFileParamsSchema>;
export type ScanContentParams = z.infer<typeof scanContentParamsSchema>;
export type ModifiedLineRange = z.infer<typeof modifiedLineRangeSchema>;

File: packages/rust-core/crates/tscanner_types/src/params.rs
use crate::enums::AiExecutionMode;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanParams {
    pub root: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config_dir: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub staged: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ai_mode: Option<AiExecutionMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub no_cache: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanFileParams {
    pub root: String,
    pub file: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanContentParams {
    pub root: String,
    pub file: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config_dir: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uncommitted: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModifiedLineRange {
    pub start_line: usize,
    pub line_count: usize,
}
